File: rag.py
# rag.py
import json
import os
import google.generativeai as genai

GOOGLE_API_KEY = os.getenv("GOOGLE_API_KEY", "")
genai.configure(api_key=GOOGLE_API_KEY)

def carregar_base():
    caminho = os.path.join("data", "nasa_dataset.json")
    try:
        with open(caminho, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception as e:
        print(f"DEBUG: Erro ao carregar base: {e}")
        return []

def gerar_resposta(prompt: str) -> str:
    try:
        data = carregar_base()
        relacionados = [
            item["description"]
            for item in data
            if prompt.lower() in item["description"].lower()
        ][:3]
        contexto = "\n".join(relacionados) if relacionados else "Nenhum dado relevante encontrado."

        entrada = f"Base de dados:\n{contexto}\n\nPergunta: {prompt}"

        # Configura o modelo Gemini
        model = genai.GenerativeModel("gemini-1.5-flash")
        response = model.generate_content(
            entrada,
            generation_config={
                "max_output_tokens": 200,
                "temperature": 0.7
            }
        )
        print(f"DEBUG: Resposta do Gemini: {response.text}")
        return response.text or "[Erro: resposta vazia do modelo]"
    except Exception as e:
        print(f"DEBUG: Erro no gerar_resposta: {e}")
        return f"[Erro ao gerar resposta: {e}]"

File: utils/rag.ts
import fs from 'fs';
import path from 'path';
import { HfInference } from '@huggingface/inference';


const HF_TOKEN = process.env.HF_TOKEN || '';
const MODEL_ID = process.env.MODEL_ID || 'mistralai/Mistral-7B-Instruct-v0.1';


if (!HF_TOKEN) {
console.warn('AVISO: HF_TOKEN não configurado. O serviço falhará sem o token.');
}


const hf = new HfInference(HF_TOKEN);


export async function generateFromPrompt(prompt: string) {
// Carrega dataset local
const datasetPath = path.join(process.cwd(), 'data', 'nasa_dataset.json');
let data: any[] = [];
try {
const raw = fs.readFileSync(datasetPath, 'utf-8');
data = JSON.parse(raw);
} catch (err) {
console.error('Falha ao ler dataset local:', err);
// continuar com data vazia
}


// Busca simples: procura ocorrências do texto no campo description/title
const q = prompt.toLowerCase().trim();
const related = data
.filter(item => {
const text = ((item.title || '') + ' ' + (item.description || '')).toLowerCase();
return text.includes(q);
})
.slice(0, 3)
.map(i => `- ${i.title}: ${i.description}`)
.join('\n');


const systemPrompt = `Você é Aurora v7 — assistente técnico para o NASA Space Apps. Use apenas as informações fornecidas na seção 'dados' para responder quando possível. Se nada for relevante, responda de forma clara, indicando que não há dados suficientes e forneça sugestões de como obter as informações.`;


const inputs = `SYSTEM:\n${systemPrompt}\n\nDADOS RELEVANTES:\n${related || 'Nenhum dado relevante encontrado no dataset.'}\n\nPERGUNTA:\n${prompt}`;


try {
const out = await hf.textGeneration({
model: MODEL_ID,
inputs,
parameters: {
max_new_tokens: 256,
temperature: 0.2
}
});


// A resposta pode estar em diferentes campos dependendo da versão do SDK
const text = (out as any)?.generated_text || (Array.isArray(out) ? out[0]?.generated_text : undefined) || JSON.stringify(out);


return text;
} catch (err: any) {
console.error('Erro ao chamar Hugging Face:', err);
throw new Error('Falha ao gerar resposta via modelo de IA: ' + (err?.message || 'Desconhecido'));
}
}